### PR TITLE
(SIMP-10197) Transition scripts for OpenLdap to 389ds

### PR DESCRIPTION
**Pull request by jeannegreulich**
- added scripts and instructions for moving Openldap users
  to a 389ds server.

SIMP-10197 #close

### CODE REVIEW

File: share/transition_scripts/6.6.0/openldap2ds389.rb [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+      newv = [ 'groupOfNames', 'nsMemberOf', 'posixGroup']
+      v.each do |nv|
+        newv << nv
+      end
+      newattr[:objectclass] = newv.uniq
+    else
+      newattr[a] = v unless @delete_keys.member?(a)
+    end
+  end
+  @ldifs_389[newdn] = newattr
```
Comment (op-ct) [inline share/transition_scripts/6.6.0/openldap2ds389.rb:37]: From the calling code, it's sort of a surprise that this happens to @ldifs_389 here.  

As a slight tweak: if the function simply returned the converted key/value pair, like

```ruby
{newdn => newattr}
```

then the calling code could take care of merging it into `@ldifs_389`:

```ruby
@ldifs_389.merge! convert_group(<args>)
```
and  `@ldifs_389` could just be a normally-scoped variable, i.e.,

```
ldifs_389.merge! convert_group(<args>)
```

That would make everything easier to keep track of and keep the intent of the function a bit clearer

File: share/transition_scripts/6.6.0/openldap2ds389.rb [hunk 2]
```diff
@@ -0,0 +1,179 @@
+#!/opt/puppetlabs/puppet/bin/ruby
```
Comment (op-ct) [inline share/transition_scripts/6.6.0/openldap2ds389.rb:1]: It's a little hard to follow this logic―it uses `@instance` variables as if they were globals, without any classes.

If the logic was wrapped in a class, the instance variables might make more sense and we could spec test the script to demonstrate that the logic works.

File: share/transition_scripts/6.6.0/openldap2ds389.rb [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        end
+      end
+      newattr[:objectclass] = newv.uniq
+    when :pwdaccountlockedtime
+      newattr[:nsaccountlock] = 'true'
+    else
+      newattr[a] = v unless @delete_keys.include?(a)
+    end
+  end
+  @ldifs_389[dn] = newattr
```
Comment (op-ct) [inline share/transition_scripts/6.6.0/openldap2ds389.rb:70]: Similar situation to `convert_group` wrt `@ldifs_389`

**Review by op-ct — CHANGES_REQUESTED** [review]
1. We really shouldn't need to vendor the `net-ldap` gem binary in `simp-utils`. The **rubygem-net-ldap.noarch** RPM provides `NET::LDAP` and is distributed with SIMP.

2. The README.md markdown formatting is wonky (but a few ``` blocks should fix it)

3. The `openldap2ds389.rb` script mutates instance variables like globals and is hard to follow. 
 The logic should have spec tests to demonstrate that it works correctly.

**Comment by op-ct** [comment]
Should the `convert_inet_org.rb` script from [SIMP-9724] be updated/deprecated along with these transition scripts?

**Review by trevor-vaughan — DISMISSED** [review]
It looks like `memberUid` needs to be changed to `member` as well

**Comment by lnemsick-simp** [comment]
Looks like the comments on the README.md were erroneously marked as resolved by GitHub.  Will address them.

**Review by op-ct — CHANGES_REQUESTED** [review]
Two of the `import_to_389ds.sh` conditionals don't work as intended, which makes me want test coverage of `openldap_to_389ds.rb` (as there is with `unpack_dvd.rb`) to verify that script logic behaves the way what we expect under the various conditions it's intended to handle.

File: share/transition_scripts/openldap_to_389ds/openldap_to_389ds.rb [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  newattrs
+end
+
+input_ldif = 'simp_openldap.ldif'
+output_ldif = 'simp_389ds.ldif'
+basedn = nil
+
+optparse = OptionParser.new do |opts|
+  opts.banner = "Usage: #{$PROGRAM_NAME} -i <inputfile> -o <outputfile> -b <base dn>"
+  opts.separator <<~HELP_MSG
```
Comment (lnemsick-simp) [inline share/transition_scripts/openldap_to_389ds/openldap_to_389ds.rb:86]: `<<~` is not recognized with the system Ruby (2.0.0) on EL7.

File: share/transition_scripts/openldap_to_389ds/openldap_to_389ds.rb [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  end
+  opts.on('-o', '--output FILE', 'file to output the ldifs to import into 389-DS server',
+          'Default: ./data/simp_389ds.ldif') do |f|
+    output_ldif = f.strip
+  end
+  opts.on('-h', '--help', 'Help') do
+    puts opts
+    exit
+  end
+  opts.separator <<~HELP_MSG
```
Comment (lnemsick-simp) [inline share/transition_scripts/openldap_to_389ds/openldap_to_389ds.rb:108]: `<<~' is not recognized by system Ruby on EL7

**Review by lnemsick-simp — DISMISSED** [review]
Instructions assume using system Ruby as the installation of net-ldap will install the gem into system Ruby.  However, the code uses syntax `<<~` which is invalid in system Ruby on EL7 (2.0.0).  Small fix to use `<<-` instead is required.

**Comment by lnemsick-simp** [comment]
working on acceptance tests for this

**Comment by lnemsick-simp** [comment]
Work still in progress because `dsidm` won't list the imported users, eventhough I can see they are there with `ldapsearch`

**Review by trevor-vaughan — APPROVED** [review]

File: share/transition_scripts/openldap_to_389ds/openldap_to_389ds.rb [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+require 'optparse'
+
+# Attributes that will cause issues in 389-DS
+# - memberuid is copied into to member before deletion
+# - sshpublickey is copied into nssshpublickey before deletion
+
+DELETE_KEYS = %i[
+  entrycsn
+  entryuuid
+  memberuid
```
Comment (jeannegreulich) [inline share/transition_scripts/openldap_to_389ds/openldap_to_389ds.rb:32]: In sssd we have the schema set to rfc2307.  This looks for group membership using memberUid.  If we do this then we need to change the schema in the ldap provider to rfc2307bis.      I don't know what other effects this might have.   If we decide to keep it rfc2307 then we need to update the docs in simp-doc.

Comment (trevor-vaughan) [reply share/transition_scripts/openldap_to_389ds/openldap_to_389ds.rb:32]: @jeannegreulich :+1: we should definitely change it to `rfc2307bis` in `simp::sssd::client`
